packages/ui/src/components/common/TokenSelector/PaymentMethod.tsx

```diff
@@ -0,0 +1,1342 @@
+      term: !isSearchTermValidAddress ? debouncedTokenSearchValue : undefined,
+      defaultList: useDefaultTokenList && !depositAddressOnly,
+      limit: 12,
+      depositAddressOnly,
+      referrer: relayClient?.source
+    }
+  )
+
+  const { data: externalTokenList, isLoading: isLoadingExternalList } =
+    useTokenList(
+      relayClient?.baseApiUrl,
+      {
+        chainIds: chainFilter.id
+          ? [chainFilter.id]
+          : configuredChains.map((c) => c.id),
+        address: isSearchTermValidAddress
+          ? debouncedTokenSearchValue
+          : undefined,
+        term: !isSearchTermValidAddress ? debouncedTokenSearchValue : undefined,
+        defaultList: false,
+        limit: 12,
+        useExternalSearch: true,
+        referrer: relayClient?.source
+      },
+      {
+        enabled: !!debouncedTokenSearchValue && !depositAddressOnly
+      }
+    )
+
+  // Merge token lists when searching
+  const combinedTokenList = useMemo(() => {
+    if (!debouncedTokenSearchValue) return tokenList
+    return mergeTokenLists([tokenList, externalTokenList])
+  }, [tokenList, externalTokenList, debouncedTokenSearchValue])
+
+  const sortedUserTokens = useEnhancedTokensList(
+    userTokens,
+    tokenBalances,
+    context,
+    multiWalletSupportEnabled,
+    chainFilter.id,
+    true
+  )
+
+  const sortedTrendingTokens = useEnhancedTokensList(
+    trendingTokens,
+    tokenBalances,
+    'to',
+    multiWalletSupportEnabled,
+    undefined,
+    false
+  )
+  const sortedCombinedTokens = useEnhancedTokensList(
+    combinedTokenList,
+    tokenBalances,
+    context,
+    multiWalletSupportEnabled,
+    chainFilter.id,
+    false
+  )
+
+  const [chainSearchInputElement, setChainSearchInputElement] =
+    useState<HTMLInputElement | null>(null)
+  const [tokenSearchInputElement, setTokenSearchInputElement] =
+    useState<HTMLInputElement | null>(null)
+
+  const inputElement = hasMultipleConfiguredChainIds
+    ? chainSearchInputElement
+    : tokenSearchInputElement
+
+  const handleChainStarToggle = useCallback(() => {
+    setStarredChainIds(getStarredChainIds())
+  }, [])
+
+  useEffect(() => {
+    if (!autoSelectToken) {
+      return
+    }
+
+    if (
+      address &&
+      isValidAddress &&
+      !token &&
+      sortedUserTokens &&
+      sortedUserTokens.length > 0
+    ) {
+      const topToken = sortedUserTokens[0]
+      setToken(topToken)
+    }
+  }, [
+    address,
+    isValidAddress,
+    token,
+    sortedUserTokens,
+    setToken,
+    autoSelectToken
+  ])
+
+  useEffect(() => {
+    if (open) {
+      setStarredChainIds(getStarredChainIds())
+    }
+  }, [open])
+
+  const resetState = useCallback(() => {
+    setTokenSearchInput('')
+    setChainSearchInputElement(null)
+    setTokenSearchInputElement(null)
+  }, [])
+
+  const triggerWithBalances = useMemo(() => {
+    const injectBalances = (node: ReactNode): ReactNode => {
```

Comment:
This is kind of funky, can you explain what's happening here and why we took this approach?